### PR TITLE
Hellomsg

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
### Purpose

Adds a pre-protocol Hello message which does not require any authorization.
The message can be of variable size, but marshalled message sent must be padded to have the correct size.
Expose device name in the connect prompts.

### Testing

Ran this on two machines, connects fine, names show up in prompt and in edit device dialog.

### CODE REVIEW

**Comment by calmh** [comment]
Looks good, but:

- Why the padding to max size? It's quite large, and different from what we do everywhere else. 
- Why remove the options field from the clusterconfig?
- What's the mysterious payload field?

**Comment by AudriusButkevicius** [comment]
1. We spoke about it being of a fixed size, hence why I went with the padding approach, but I agree we could have magic + len + msg instead, though we'd still have to validate the len to make sure it's not more than something.
2. 
The realignment deceives you:
```diff
-	Folders       []Folder // max:1000000
 -	Options       []Option // max:64
 +	Folders []Folder // max:1000000
 +	Options []Option // max:64
```
3. This is for future expansion, for providing the one-off codes which allow adding folders/devices. Made it 256 bytes and fairly generic as it's not clear how much data we will need.
We might want to make it bigger as well, or allow an array, so that we could send multiple codes?

**Comment by calmh** [comment]
1. Yeah I think we misunderstood each other. I think a length word, and a hard coded max length like 1024 bytes or so where we drop the connection if exceeded would be nicer. 
2. :+1:
3. I'm unconvinced. I think we should leave this out until we need it, and then add whatever it is we need of the correct type at the time.

**Comment by AudriusButkevicius** [comment]
The problem with #3 is that we are potentially again introducing a protocol incompatibility some time later.

**Comment by calmh** [comment]
Not necessarily. If we can handle a zero length payload field, we can also handle the absence of one. Likewise if we can ignore data in a payload field that we don't look at, we can also ignore data after the XDR structure we expect.

File: lib/protocol/message.go [hunk 1]
```diff
@@ -13,13 +13,8 @@ import (
...
 var (
 	sha256OfEmptyBlock        = sha256.Sum256(make([]byte, BlockSize))
-	Version             int32 = 1
-	HelloMessageMaxSize       = 4 + 64 + // DeviceName (4 bytes for length + 64 max length)
-		4 + 64 + // ClientName
-		4 + 64 + // ClientVersion
-		4 + // ProtocolVersion
-		4 + // PayloadType
-		4 + 256 // Payload
+	Version            int32  = 1
```
Comment (calmh) [inline lib/protocol/message.go]: Hmm. We have a version field in the header of every message, and it has a different value (version=0 currently). What's the relation between this version and that?

**Comment by AudriusButkevicius** [comment]
If we bump that version (which for some reason we don't), it will just throw up with XDR errors, and not a meaningful error.

**Comment by calmh** [comment]
No it won't, it'll cause a protocol mismatch error. And you know perfectly well why we're not bumping it all the time. :)

**Comment by AudriusButkevicius** [comment]
We should start doing it then, anyways removed it.

File: lib/connections/connections.go [hunk 2]
```diff
@@ -472,6 +476,54 @@ func (s *connectionService) CommitConfiguration(from, to config.Configuration) b
 	return true
 }
 
+func exchangeHello(c net.Conn, h protocol.HelloMessage) (protocol.HelloMessage, error) {
+	if err := c.SetDeadline(time.Now().Add(2 * time.Second)); err != nil {
+		return protocol.HelloMessage{}, err
+	}
+	defer c.SetDeadline(time.Time{})
```
Comment (calmh) [inline lib/connections/connections.go]: Neat use of defer with the deadline by the way

**Comment by calmh** [comment]
:+1:

**Comment by calmh** [comment]
@st-review: merge this please

protocol: Add "Hello" message at connection start, also for unauthed peers

**Comment by st-review** [comment]
Merge failed:

```
$ git fetch -f origin refs/pull/2875/head:pr-2875
$ git fetch -f origin master:orig/master
From github.com:syncthing/syncthing
   29913dd..6a3f3f5  master     -> orig/master
$ git reset --hard
HEAD is now at 6a3f3f5 gui: Add theme.css, move dark theme, adjust popover advanced folder settings colors (fixes #2878)
$ git checkout master
Already on 'master'
Your branch is up-to-date with 'origin/master'.
$ git reset --hard orig/master
HEAD is now at 6a3f3f5 gui: Add theme.css, move dark theme, adjust popover advanced folder settings colors (fixes #2878)
$ git clean -fxd
$ git merge --squash --no-commit pr-2875
Auto-merging lib/connections/connections.go
CONFLICT (content): Merge conflict in lib/connections/connections.go
Auto-merging lib/auto/gui.files.go
CONFLICT (content): Merge conflict in lib/auto/gui.files.go
Auto-merging gui/default/index.html
Squash commit -- not updating HEAD
Automatic merge failed; fix conflicts and then commit the result.

```

**Comment by calmh** [comment]
Needs a merge of master into this branch to resolve those conflicts, then good to go

**Comment by AudriusButkevicius** [comment]
@st-review: merge this please

protocol: Add "Hello" message at connection start, also for unauthed peers

**Comment by st-review** [comment]
OK, merged as b4f9a55e6eff37e891a26909de7a378014e64cd0. Thanks, @AudriusButkevicius!

**Pull request closed without merging**